src/main/markdown.ts

```diff
@@ -0,0 +1,128 @@
+import { getGlobalRouteRegistry, RouteDefinition } from './router';
+import { groupBy } from './util';
+
+export function generateMarkdownApiSpec() {
+    const doc: string[] = [];
+    const routes = getGlobalRouteRegistry().filter(_ => !_.isMiddleware);
+    const groupByPath = groupBy(routes, r => r.path);
+    for (const [_path, routes] of groupByPath) {
+        for (const route of routes) {
+            doc.push(...generateEndpointSpec(route));
+        }
+    }
+    return doc;
+}
+
+function generateEndpointSpec(endpoint: RouteDefinition): string[] {
+    const method = endpoint.method.toUpperCase();
+    let header = `## ${method} ${endpoint.path}`;
+    if (endpoint.deprecated) {
+        header += ' **(deprecated)**';
+    }
+
+    const { queryParams, bodyParams } = generateParamSpec(endpoint);
+    const responses = generateResponsesSpec(endpoint);
+
+    const spec: string [] = [
+        '\n',
+        header,
+        endpoint.summary.trim(),
+    ];
+
+    if (queryParams.length > 0) {
+        spec.push('\n', '### Query Params', ...queryParams);
+    }
+
+    if (bodyParams.length > 0) {
+        spec.push('\n', '### Body Params', ...bodyParams);
+    }
+
+    if (responses.length > 0) {
+        spec.push('\n', '### Responses', ...responses);
+    }
+
+    return spec;
+}
+
+function generateParamSpec(endpoint: RouteDefinition) {
+    const bodyParams: string[] = [];
+    const queryParams: string[] = [];
+
+    for (const p of endpoint.params) {
+        let header = '- ' + p.name;
+
+        if (p.required) {
+            header += ' (*required*)';
```

Comment:
I think I'd actually prefer dropping this and adding "optional" instead (since optional params are less frequent).
Overall I think params can be a bit tidier, e.g.

```
- limit: number (optional, default 100, max 1000) — how many records to return per request
```

These are just thoughts ofc. Let's see if anyone else has any other opinions?

Reply:
yup, that optional make sense and dropping `type` will also make it tidier.

About putting params inline, it looks good, but we. might need to think a bit more about the array. above will work well for non-object items:
```
- tags: array (items: string, default: [])
``` 

but when the item's type is object, we probably can do like this:
- data: array (items: object, default: [])
  - items: object
     - properties:
       - id: string
       - proxyConfigId: string

Reply:
Agreed. Better yet, for non-primitive types we can just emit JSON schema as code block.